### PR TITLE
PANDA 3204, Update UT for stock notification.

### PR DESCRIPTION
**Pull request by Terry01Tan**

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 1]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
-      },
-      {
-        channel: 'SMS',
-        value: '13800000831',
-      },
-    ];
-    component.subscribeSuccess$ = of(true);
-    component.selectedChannels = selectedChannels;
+  it('should show notification dialog', () => {
     fixture.detectChanges();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: if you use 'fixture.detectChanges();' in  beforeEach segment, you do not need to call it in this place

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 2]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
+
+    expect(el.query(By.css('.cx-spinner'))).toBeTruthy();
   });
 
-  it('should show loading template when data is not ready', () => {
-    selectedChannels = [];
-    component.selectedChannels = selectedChannels;
-    component.subscribeSuccess$ = of(false);
+  it('should be able to close dialog', () => {
     fixture.detectChanges();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: ditto

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 3]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
-    const notificationPreference = fixture.debugElement.queryAll(
-      By.css('.cx-notification-preference-span')
-    );
-    expect(notificationPreference.length).toEqual(0);
+
+    el.query(By.css('[data-test]="close-button"')).nativeElement.click();
+    expect(ngbActiveModal.dismiss).toHaveBeenCalled();
   });
 
   it('should be able to close dialog', () => {
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: The name of this case is the same as before, please merge them or provide different name

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 4]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
-    component.subscribeSuccess$ = of(true);
     fixture.detectChanges();
-    fixture.debugElement.queryAll(By.css('button'))[0].nativeElement.click();
-    expect(ngbActiveModal.dismiss).toHaveBeenCalledWith('Cross click');
+
+    el.query(By.css('[data-test]="okay-button"')).nativeElement.click();
+    expect(ngbActiveModal.dismiss).toHaveBeenCalled();
+  });
+
+  it('should be able to unsubscribe/rest in destory', () => {
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: I don't see any 'unsubscribe' expected

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 5]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
     expect(component).toBeTruthy();
   });
 
-  it('should not load stock notificaiton data, when customer is not login', () => {
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: Personal view, the order of the cases can be:
 - Test the display of elements when customer is login/not login, with/without channel set ....
 - Test operation such as create stock notification(loading/success/fail), remove stock notification(loading/success/fail)

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 6]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
-      preferences: [],
-    };
-    userService.getNotificationPreferences.and.returnValue(
-      of(emptyNotificationPreferenceList)
+      cold('-b|', { a: true })
     );
     fixture.detectChanges();
+    const button = el.query(By.css('[data-test="stocknotify-button"]'))
+      .nativeElement;
+    button.click();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: Maybe you need to call fixture.detectChanges(); after call  button.click()

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 7]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
-      'stockNotification.stopNotify'
-    );
+    const button = el.query(By.css('[data-test="stocknotify-button"]'))
+      .nativeElement;
+    button.click();
+    fixture.detectChanges();
+
+    expect(productInterestService.createBackInStock).toHaveBeenCalled();
+    expect(ngbModal.open).toHaveBeenCalled();
+    expect(globalMessageService.add).not.toHaveBeenCalled();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: Why the name is 'show dialog and global message' but you expect(globalMessageService.add).not.toHaveBeenCalled();  ?

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 8]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
+  it('should show loading when delete stock notification', () => {
+    productInterestService.getBackInStockSubscribed.and.returnValue(of(true));
+    productInterestService.getDeleteBackInStockLoading.and.returnValue(
+      cold('-a|', { a: true })
     );
-  });
+    fixture.detectChanges();
+    const button = el.query(By.css('[data-test="stocknotify-button"]'))
+      .nativeElement;
+    button.click();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: ditto, call fixture.detectChanges(); after calling click()

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 9]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
-    );
-
-    userService.getNotificationPreferences.and.returnValue(
-      of(basicNotificationPreferenceList)
+      cold('-b|', { a: true })
     );
     fixture.detectChanges();
+    const button = el.query(By.css('[data-test="stocknotify-button"]'))
+      .nativeElement;
+    button.click();
```
Comment (Qingyu-Yang) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: ditto

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 10]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
-    expect(plist.length).toEqual(6);
-    expect(plist[1].nativeElement.textContent).toContain('EMAIL: test@sap.com');
-    expect(plist[2].nativeElement.textContent).toContain('SMS: 13800000831');
+
+    expect(el.query(By.css('[data-test]="notifydialog-title"'))).toBeTruthy();
+    expect(el.query(By.css('[data-test]="close-button"'))).toBeTruthy();
+    expect(el.query(By.css('[data-test]="notify-channel"'))).toBeTruthy();
+    expect(
+      el.queryAll(By.css('[data-test]="notifydialog-link"')).length
+    ).toEqual(2);
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: there is no mock for selected channels

Comment (kate1zheng) [reply projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: there are two links, settings and my interests should be verified separately

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 11]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
+    expect(el.query(By.css('[data-test]="ok-button"'))).toBeTruthy();
   });
 
-  it('should show manage channels link and manage subscriptions link', () => {
-    component.subscribeSuccess$ = of(true);
+  it('should show spinner when loading', () => {
+    component.subscribeSuccess$ = of(false);
+    productInterestService.createBackInStock.and.returnValue({});
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: why we need mock createBackInStock ?

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 12]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
+    expect(ngbActiveModal.dismiss).toHaveBeenCalled();
   });
 
   it('should be able to close dialog', () => {
-    component.subscribeSuccess$ = of(true);
     fixture.detectChanges();
-    fixture.debugElement.queryAll(By.css('button'))[0].nativeElement.click();
-    expect(ngbActiveModal.dismiss).toHaveBeenCalledWith('Cross click');
+
+    el.query(By.css('[data-test]="okay-button"')).nativeElement.click();
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: this okay-button is the same with the button ok-button? why use different name?

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts [hunk 13]
```diff
@@ -59,61 +59,53 @@ describe('NotificationDialogComponent', () => {
...
+    expect(ngbActiveModal.dismiss).toHaveBeenCalled();
+  });
+
+  it('should be able to unsubscribe/rest in destory', () => {
+    fixture.detectChanges();
+
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/notification-dialog/notification-dialog.component.spec.ts]: there is no need to call detectChanges()
the description should be corrected, what does '/rest' mean? and there is a typo for destroy

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 14]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
     fixture.detectChanges();
-    component.logged$.subscribe(x => expect(x).toBeFalsy());
-  });
 
-  it('should load stock notification data , when customer is login', () => {
-    authService.getUserToken.and.returnValue(of(userToken));
-    fixture.detectChanges();
-    component.logged$.subscribe(x => expect(x).toBeTruthy());
+    expect(
+      el.query(By.css('[data-test]="stocknotify-link"')).nativeElement
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: the description should change to 'show disabled button and setting link for login customer without setting notification channel'
the element naming should be meaningful, don't alway use stocknotify as prefix

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts (hunk 5 above)
Comment (kate1zheng) [reply projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: agree!!

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts (hunk 7 above)
Comment (kate1zheng) [reply projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: is there any message shown when subscribed the notification???

Comment (kate1zheng) [reply projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: we don't need to verify there is no message to shown

File: projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts [hunk 15]
```diff
@@ -183,129 +154,108 @@ describe('StockNotificationComponent', () => {
...
+    ).toBeTruthy();
+    expect(
+      el.query(By.css('[data-test]="stocknotify-note"')).nativeElement
+    ).toBeTruthy();
+    expect(
+      el.query(By.css('[data-test]="stocknotify-button"')).nativeElement
```
Comment (kate1zheng) [inline projects/storefrontlib/src/cms-components/product/product-interests/stock-notification/stock-notification.component.spec.ts]: we should differentiate which button - notify me /do not notify me

**Pull request merged**